### ISSUE TITLE
How to resume YOLOv8 hyperparameter tuning?

### ISSUE THREAD

**Issue by inmess** [ultralytics/ultralytics#5892]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I was tuning yolov8n with my data. But somehow it just get stuck, and I really did not find any method to resume it, not with `resume=True` anyway. Is the only way to restore it to restart a new tuning by the best hyperparameters produced before?

Code I used:
```python
from ultralytics import YOLO

model = YOLO('tune.pt')

model.tune(data='cc_data.yaml', epochs=30, iterations=500, batch=150, imgsz=640)
```

And it stopped in the middle without any information:
![image](https://github.com/ultralytics/ultralytics/assets/48652450/f853dd9b-b6c9-4762-97cb-8863c2a7096e)

**Comment by Pzzzzzzzzz** [comment]
your batch is too large

**Comment by glenn-jocher** [comment]
@Pzzzzzzzzz hi there! The error you're encountering indicates that the size of the batch of images you're using to train the YOLOv8 model might be too large for your available computing resources, particularly your GPU memory.

Batch size refers to the number of images that the model processes at once during training. If the batch size is too large, it can exceed your GPU's memory capacity, causing a failure in your training process.

Try reducing the batch size in your training configuration. This will decrease the amount of memory required for each training step, preventing an overflow of your GPU memory. Though this might slow down the training process, it should solve your problem and let you proceed.

Always remember that selecting the right batch size depends on your specific hardware setup and the complexity of your model. It's a balance between the computation speed and the available memory.

Hope this helps! Let me know if you have any other questions.

**Comment by inmess** [comment]
> your batch is too large

@Pzzzzzzzzz @glenn-jocher Actually, I got 24G GPU Memory, my GPU is still not fully used (I was trying to use GPU as much as possible). Is the batch size still too large here?  And as you can see in the image I showed above, the tuning stopped in the 71st train process, not the 70 times of training before it, with no error message presenting.

**Comment by glenn-jocher** [comment]
Hi @inmess,

Thank you for providing additional details. While it's indeed beneficial to use your GPU resources to their fullest, it's also critical to ensure that the batch size isn't too large to hinder the training process. Your GPU memory isn't just used for holding the batch of images, but it also stores intermediate computations, model parameters, and gradients which can vary during the training process.

Therefore, even if it seems like your GPU memory isn't fully used initially, it might spike up in usage at certain iterations causing the process to stop. This could explain why your tuning stopped at the 71st train process without clear error messages: the system may have silently run out of GPU memory.

Even with 24GB of GPU memory, a very large batch size could cause unpredictable behaviour. I would recommend trying to reduce the batch size slightly to see if the training can complete without being interrupted. This would also help verify if the issue is indeed because of the batch size being too large. Your training speed will be slower, but it will likely be more stable.

I hope that helps. Let's continue to troubleshoot this together. Please let me know how the training goes with a smaller batch size.

**Comment by inmess** [comment]
> I hope that helps. Let's continue to troubleshoot this together. Please let me know how the training goes with a smaller batch size.

@glenn-jocher Appreciate, I'm now tuning with a smaller batch size (100), will see if this works in about 40hrs. But I'm still wondering if there is a method to continue the tuning process that haven't finished?

**Comment by glenn-jocher** [comment]
Hi @inmess, 

Great to hear that you've started the tuning process with a smaller batch size. I look forward to hearing how that goes!

As far as your question about resuming an unfinished tuning process, typically in most machine learning workflows, if the process stops unexpectedly it can't be resumed from the middle. You would have to start the process again, possibly with the adjusted parameters if they were the cause of the problem, like possibly reducing the batch size as we discussed.

The reasons for this have to do with the mechanics of hyperparameter tuning: the tuning process uses the results of previous iterations to decide on the parameters for the next iteration. If the process is stopped midway, the model loses this context and so a fresh run is required to maintain the integrity of the results.

That being said, I agree that it might be beneficial to have a feature that saves the state of the tuning process after each iteration, allowing for a restart from the last completed iteration in case of interruptions. We'll take this into consideration for the future. Thank you for your feedback and please let us know how your tuning goes!

**Comment by inmess** [comment]
Thanks @glenn-jocher ,

The tuning have the same problem after I gave a smaller batch size. 

![image](https://github.com/ultralytics/ultralytics/assets/48652450/ea97bfdd-f160-4c49-8b48-e1724e3c6db2)

This time stopped in 109/500 iterations. Do you have any idea why would this happen except a large batch size? I mean I can give another smaller batch size to it, but but this cost a huge amount of time, and I cannot stay in front of the monitor all the day as well. 

> the tuning process uses the results of previous iterations to decide on the parameters for the next iteration.

As you mentioned, this would be really helpful if there is any method to figure out the reason it stopped, to have a complete tuning process. Btw I don't think this is due to a large resource cost, cuz it's like freeze here, meanwhile the system resources are pretty much not used. ⬇️
![image](https://github.com/ultralytics/ultralytics/assets/48652450/3dc59015-eb65-495b-ae73-6255883b6141)
![image](https://github.com/ultralytics/ultralytics/assets/48652450/54cd7e3e-ae43-4a1c-af21-e6c39257c240)

**Comment by glenn-jocher** [comment]
Hi @inmess,

Thanks for sharing the results. I'm sorry to hear the tuning process is still experiencing issues even after reducing the batch size. Based on your observations, it does indeed seem like resource exhaustion isn't the issue here.

While it's hard to tell without further investigation, other potential causes for the freeze could be:
- An unexpected bug or error in the dataset or model code: This could silently disrupt the model's training process without triggering a visible error.
- External system issue: Things like abrupt power loss, sudden system reboots, or other software-related issues could cause the process to terminate prematurely.

To get more details on what's happening, you could consider adding more logging or monitoring to the training process which could help highlight when and possibly why the freeze is occurring. If there's an issue with the dataset or model code, these logs might help identify it.

You also mentioned the desire for continuation of the process after a sudden stop. While it's generally a challenging proposition to resume an interrupted hyperparameter tuning run from where it left off, it can definitely be a boon in long-running processes like these. For now, I would recommend adding checkpoints in your code, which periodically save out the status of the model. This way, in case of an unexpected failure, you could at least have the option to "manually" resume the process from the last saved checkpoint.

I understand this situation can be quite frustrating, especially considering the amount of time invested in these training runs. Let's continue to solve this issue together.

**Comment by finlaymorrison** [comment]
This is a feature which I am very interested in --- for my use case I have job times limited to 60 hours and so I cant do a great deal of tuning in a single run. I am working on a solution to this which adds another argument to the tuner which is the filepath to an output csv file generated in a previous tuning run, where the contents of this csv file are copied over to the csv file generated by the new training run.

If this is a workable solution, should I create a new issue about it to link a PR to, or should I create a PR and link to this issue?

**Comment by glenn-jocher** [comment]
Hello @finlaymorrison,

Your idea of a feature that enables the continuation of tuning from a previous run sounds valuable, particularly for scenarios involving time-constrained environments. The approach of utilizing a CSV file to preserve the state of the previous run and feed it into the new one is quite pragmatic.

Yes, I would suggest creating a new issue for this proposed feature on the repository. This gives your idea visibility and provides a structured platform for other contributors and maintainers to discuss its implementation details. Once you have the issue created and there is consensus on the approach, you can then create a Pull Request (PR) and link it back to that issue.

Remember to provide thorough details in the issue documentation—like the problem, the proposed solution, and how it can integrate with the existing framework. Once the issue is in place and you are ready, you may proceed with the PR, referencing the newly created issue.

Contributions to the project are always welcome, and I appreciate your involvement in making the YOLOv8 framework better for everyone. Looking forward to your contribution!

**Comment by unrue** [comment]
Apart batch size problem, how to resume hyperparameter tuning on yolov8 ?

**Comment by finlaymorrison** [comment]
@unrue the feature does not currently exist, but if you look at the linked PR #6699 , you will see that this feature has been implemented, it is just waiting on being merged.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #20037](https://github.com/ultralytics/ultralytics/pull/20037). This update introduces exciting new functionality and improvements, including the ability to resume interrupted hyperparameter tuning sessions with the new `resume=True` option in `model.tune()`. This works seamlessly for both standard and Ray Tune workflows, helping save time and resources by continuing from the last interrupted session instead of starting over. The PR also includes updated documentation and internal logic enhancements to support this feature. 

Credit goes to @Y-T-G for the implementation and @glenn-jocher for contributions to this update. The version has been bumped to `8.3.103`, so you can test these changes by updating to the latest version with:

```
pip install -U ultralytics>=8.3.103
```

Alternatively, you can install directly from the main branch:

```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Feel free to test and let us know if this resolves the issue for you. If the issue persists or you encounter any other challenges, additional feedback is always welcome. Thank you 🙏 for bringing this to our attention and helping improve Ultralytics!
